aiidalab_widgets_base/viewers.py

```diff
@@ -222,7 +228,91 @@ def _download_tab(self):
         self.screenshot_btn.on_click(lambda _: self._viewer.download_image())
         self.screenshot_box = ipw.VBox(children=[ipw.Label("Create a screenshot:"), self.screenshot_btn])
 
-        return ipw.VBox([self.download_box, self.screenshot_box])
+        # 4. Render a high quality image
+        self.render_btn = ipw.Button(description="Render", icon='fa-paint-brush')
+        self.render_btn.on_click(self._render_structure)
+        self.render_box = ipw.VBox(children=[ipw.Label("Render a image with POVRAY:"), self.render_btn])
+
+        return ipw.VBox([self.download_box, self.screenshot_box, self.render_box])
+
+    def _render_structure(self, change=None):
+        """Render the structure with POVRAY."""
+
+        if not isinstance(self.structure, Atoms):
+            return 
+
+        self.render_btn.disabled = True 
+        omat = np.array(self._viewer._camera_orientation)
+        omat = omat.reshape(4, 4)
+        omat = np.transpose(omat)
+
+        zfactor = norm(omat[0, 0:3])
+        omat[0:3, 0:3] = omat[0:3, 0:3]/zfactor
+
+        bb = deepcopy(self.structure)
+
+        for i in bb:
+            ixyz = np.array([i.x, i.y, i.z])
+            ixyz = ixyz + omat[0:3, 3];
+            ixyz = omat[0:3, 0:3].dot(ixyz)
+            i.x = -ixyz[0]
+            i.y =  ixyz[1]
+            i.z =  ixyz[2]
+        
+        vertices = [];
+    
+        cell = bb.get_cell()
+        vertices.append(np.array([0, 0, 0]))
+        vertices.extend(cell)
+        vertices.extend([cell[0]+cell[1], cell[0]+cell[2], cell[1]+cell[2], cell[0]+cell[1]+cell[2]])
+
+        for n, i in enumerate(vertices):
+            ixyz = omat[0:3, 0:3].dot(i + omat[0:3, 3])
+            vertices[n] = np.array([-ixyz[0], ixyz[1], ixyz[2]])
+    
+        camera = Camera('location', [0, 0, -zfactor/1.5], 'look_at', [0.0, 0.0, 0.0])
+        light = LightSource([0, 0, -100.0], 'color',  [1.5, 1.5, 1.5])
+
+        spheres = [Sphere( [i.x, i.y, i.z], Radius[i.symbol], 
+                            Texture(Pigment( 'color', np.array(Colors[i.symbol]))), 
+                                Finish('phong', 0.9,'reflection', 0.05)) for i in bb]
+
+        bonds = []
+        for x, i in enumerate(bb):
```

Comment:
@csadorf suggests another tool for the nearest neighbours finding: https://freud.readthedocs.io/en/stable/gettingstarted/tutorial/neighborfinding.html?highlight=bond

Reply:
Certainly in favor using freud for nearest neighbor finding! We would need to install it within the environment, but that shouldn't be a big deal. Just let me know and I can create a PR for that.

Reply:
Hi @dou-du. Together with @cpignedoli we had a look at the ASE functionality, and it appears the code you need is rather simple. Just use something as follows:

```python
from ase import neighborlist
cutOff = neighborlist.natural_cutoffs(s) # Takes the cutoffs from the ASE database
neighborList = neighborlist.NeighborList(cutOff, self_interaction=False, bothways=False)
neighborList.update(s)
matrix = neighborList.get_connectivity_matrix()
```
This is fast and, more importantly, this scales linearly. The bonds can be further extracted from the `matrix.keys()`, which looks as follows:
```python
for k in matrix.keys():
    print(k)

# Content:
(0, 67)  # (index_atom1, index_atom2)
(0, 1)
(0, 66)
(1, 68)
...
```

Reply:
Computing time as the function of the number of atoms:
```
# N atoms,  time(s):
1 *   1516, 1.026095493659973
2 *   1516, 1.928958248347044
4 *   1516, 3.314179550856352
8 *   1516, 6.640921363607049
16 *  1516, 13.793273512274027
32 * 1516, 25.652117231860757
```

Reply:
> Hi @dou-du. Together with @cpignedoli we had a look at the ASE functionality, and it appears the code you need is rather simple. Just use something as follows:
> 
> ```python
> from ase import neighborlist
> cutOff = neighborlist.natural_cutoffs(bb) # Takes the cutoffs from the ASE database
> neighborList = neighborlist.NeighborList(cutOff, self_interaction=False, bothways=False)
> neighborList.update(bb)
> matrix = neighborList.get_connectivity_matrix()
> ```
> 
> This is fast and, more importantly, this scales linearly. The bonds can be further extracted from the `matrix.keys()`, which looks as follows:
> 
> ```python
> for k in matrix.keys():
>     print(k)
> 
> # Content:
> (0, 67)  # (index_atom1, index_atom2)
> (0, 1)
> (0, 66)
> (1, 68)
> ...
> ```

I had tried this. But there seems a problem for the periodic boundary condition. It plots the figure as:

![C240](https://user-images.githubusercontent.com/56022756/107988451-5717f280-6fd0-11eb-9d1b-639c218422c9.png)

Do you have a idea how to fix this ?

Reply:
well I seea t least two options:
1)check in ase neighborlist the option to avoid PBC (it exists if I remember well)
2) when attempting to draw teh bond check it's length, it is bigger than the cov radii bond length
just draw for each atom involved half of the bond in reciprocal direction

Reply:
as @cpignedoli says you might just switch off the periodic boundary conditions (PBC) prior to computing the bonds:
```python
bb.pbc = False
```
I tested it, and it seems to work.

Reply:
> well I see at least two options:
> 1)check in ase neighborlist the option to avoid PBC (it exists if I remember well)
> 2) when attempting to draw the bond check it's length, it is bigger than the cov radii bond length
> just draw for each atom involved half of the bond in reciprocal direction

Thanks a lot! I fixed it by turning off the pdc for the structures.